### # Hackathon - XYZ_Vaccine & Seasonal_Vaccine

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

In [2]:
x_train = pd.read_csv("training_set_features.csv")

In [3]:
x_train.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]

In [4]:
print("x_train.shape: {}" .format(x_train.shape))

x_train.shape: (26707, 36)


In [5]:
x_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                26707 non-null  int64  
 1   xyz_concern                  26615 non-null  float64
 2   xyz_knowledge                26591 non-null  float64
 3   behavioral_antiviral_meds    26636 non-null  float64
 4   behavioral_avoidance         26499 non-null  float64
 5   behavioral_face_mask         26688 non-null  float64
 6   behavioral_wash_hands        26665 non-null  float64
 7   behavioral_large_gatherings  26620 non-null  float64
 8   behavioral_outside_home      26625 non-null  float64
 9   behavioral_touch_face        26579 non-null  float64
 10  doctor_recc_xyz              24547 non-null  float64
 11  doctor_recc_seasonal         24547 non-null  float64
 12  chronic_med_condition        25736 non-null  float64
 13  child_under_6_mo

In [6]:
x_train["income_poverty"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 26707 entries, 0 to 26706
Series name: income_poverty
Non-Null Count  Dtype 
--------------  ----- 
22284 non-null  object
dtypes: object(1)
memory usage: 208.8+ KB


In [7]:
x_train.columns

Index(['respondent_id', 'xyz_concern', 'xyz_knowledge',
       'behavioral_antiviral_meds', 'behavioral_avoidance',
       'behavioral_face_mask', 'behavioral_wash_hands',
       'behavioral_large_gatherings', 'behavioral_outside_home',
       'behavioral_touch_face', 'doctor_recc_xyz', 'doctor_recc_seasonal',
       'chronic_med_condition', 'child_under_6_months', 'health_worker',
       'health_insurance', 'opinion_xyz_vacc_effective', 'opinion_xyz_risk',
       'opinion_xyz_sick_from_vacc', 'opinion_seas_vacc_effective',
       'opinion_seas_risk', 'opinion_seas_sick_from_vacc', 'age_group',
       'education', 'race', 'sex', 'income_poverty', 'marital_status',
       'rent_or_own', 'employment_status', 'hhs_geo_region', 'census_msa',
       'household_adults', 'household_children', 'employment_industry',
       'employment_occupation'],
      dtype='object')

In [8]:
x_test = pd.read_csv("test_set_features.csv")

In [9]:
x_test.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0          26707          2.0            2.0                        0.0   
1          26708          1.0            1.0                        0.0   
2          26709          2.0            2.0                        0.0   
3          26710          1.0            1.0                        0.0   
4          26711          3.0            1.0                        1.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   1.0                   0.0                    1.0   
1                   0.0                   0.0                    0.0   
2                   0.0                   1.0                    1.0   
3                   0.0                   0.0                    0.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          1.0                      0.0   
1                          0.0                      0.0   
2                          1.0                      1.0   
3                          0.0                      0.0   
4                          1.0                      1.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...                  > $75,000     Not Married   
1                    0.0  ...              Below Poverty     Not Married   
2                    1.0  ...                  > $75,000         Married   
3                    0.0  ...  <= $75,000, Above Poverty         Married   
4                    1.0  ...  <= $75,000, Above Poverty     Not Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0         Rent            Employed        mlyzmhmf  MSA, Not Principle  City   
1         Rent            Employed        bhuqouqj                   Non-MSA   
2          Own            Employed        lrircsnp                   Non-MSA   
3          Own  Not in Labor Force        lrircsnp  MSA, Not Principle  City   
4          Own            Employed        lzgpxyit                   Non-MSA   

   household_adults  household_children  employment_industry  \
0               1.0                 0.0             atmlpfrs   
1               3.0                 0.0             atmlpfrs   
2               1.0                 0.0             nduyfdeo   
3               1.0                 0.0                  NaN   
4               0.0                 1.0             fcxhlnwr   

   employment_occupation  
0               hfxkjkmi  
1               xqwwgdyp  
2               pvmttkik  
3                    NaN  
4               mxkfnird  

[5 rows x 36 columns]

In [10]:
x_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26708 entries, 0 to 26707
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                26708 non-null  int64  
 1   xyz_concern                  26623 non-null  float64
 2   xyz_knowledge                26586 non-null  float64
 3   behavioral_antiviral_meds    26629 non-null  float64
 4   behavioral_avoidance         26495 non-null  float64
 5   behavioral_face_mask         26689 non-null  float64
 6   behavioral_wash_hands        26668 non-null  float64
 7   behavioral_large_gatherings  26636 non-null  float64
 8   behavioral_outside_home      26626 non-null  float64
 9   behavioral_touch_face        26580 non-null  float64
 10  doctor_recc_xyz              24548 non-null  float64
 11  doctor_recc_seasonal         24548 non-null  float64
 12  chronic_med_condition        25776 non-null  float64
 13  child_under_6_mo

In [11]:
print("x_test.shape: {}" .format(x_test.shape))

x_test.shape: (26708, 36)


In [12]:
y_train = pd.read_csv("training_set_labels.csv")

In [13]:
y_train.head()

respondent_id  xyz_vaccine  seasonal_vaccine
0              0            0                 0
1              1            0                 1
2              2            0                 0
3              3            0                 1
4              4            0                 0

In [14]:
print("y_train.shape: {}" .format(y_train.shape))

y_train.shape: (26707, 3)


In [15]:
y_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   respondent_id     26707 non-null  int64
 1   xyz_vaccine       26707 non-null  int64
 2   seasonal_vaccine  26707 non-null  int64
dtypes: int64(3)
memory usage: 626.1 KB


In [16]:
df = pd.concat([x_train, x_test], axis = 0)

In [17]:
df.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]

In [18]:
print("df.shape: {}" .format(df.shape))

df.shape: (53415, 36)


In [19]:
df.reset_index(drop = True, inplace = True)

### # Step-01: Check Missing Values

In [20]:
df.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]

In [21]:
df.tail()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
53410          53410          1.0            1.0                        0.0   
53411          53411          3.0            1.0                        0.0   
53412          53412          0.0            1.0                        0.0   
53413          53413          3.0            1.0                        0.0   
53414          53414          2.0            1.0                        0.0   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
53410                   1.0                   0.0                    1.0   
53411                   1.0                   0.0                    1.0   
53412                   0.0                   0.0                    0.0   
53413                   1.0                   0.0                    1.0   
53414                   0.0                   0.0                    1.0   

       behavioral_large_gatherings  behavioral_outside_home  \
53410                          0.0                      0.0   
53411                          1.0                      1.0   
53412                          0.0                      0.0   
53413                          0.0                      1.0   
53414                          0.0                      0.0   

       behavioral_touch_face  ...             income_poverty  marital_status  \
53410                    1.0  ...                        NaN             NaN   
53411                    1.0  ...              Below Poverty         Married   
53412                    0.0  ...              Below Poverty     Not Married   
53413                    0.0  ...  <= $75,000, Above Poverty         Married   
53414                    1.0  ...                        NaN     Not Married   

       rent_or_own   employment_status  hhs_geo_region  \
53410          NaN                 NaN        dqpwygqj   
53411         Rent            Employed        qufhixun   
53412         Rent  Not in Labor Force        qufhixun   
53413          Own  Not in Labor Force        bhuqouqj   
53414         Rent            Employed        lrircsnp   

                     census_msa  household_adults  household_children  \
53410       MSA, Principle City               1.0                 1.0   
53411                   Non-MSA               1.0                 3.0   
53412  MSA, Not Principle  City               1.0                 0.0   
53413  MSA, Not Principle  City               1.0                 0.0   
53414       MSA, Principle City               0.0                 0.0   

       employment_industry  employment_occupation  
53410                  NaN                    NaN  
53411             fcxhlnwr               vlluhbov  
53412                  NaN                    NaN  
53413                  NaN                    NaN  
53414                  NaN               xtkaffoo  

[5 rows x 36 columns]

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53415 entries, 0 to 53414
Data columns (total 36 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                53415 non-null  int64  
 1   xyz_concern                  53238 non-null  float64
 2   xyz_knowledge                53177 non-null  float64
 3   behavioral_antiviral_meds    53265 non-null  float64
 4   behavioral_avoidance         52994 non-null  float64
 5   behavioral_face_mask         53377 non-null  float64
 6   behavioral_wash_hands        53333 non-null  float64
 7   behavioral_large_gatherings  53256 non-null  float64
 8   behavioral_outside_home      53251 non-null  float64
 9   behavioral_touch_face        53159 non-null  float64
 10  doctor_recc_xyz              49095 non-null  float64
 11  doctor_recc_seasonal         49095 non-null  float64
 12  chronic_med_condition        51512 non-null  float64
 13  child_under_6_mo

In [23]:
df.columns

Index(['respondent_id', 'xyz_concern', 'xyz_knowledge',
       'behavioral_antiviral_meds', 'behavioral_avoidance',
       'behavioral_face_mask', 'behavioral_wash_hands',
       'behavioral_large_gatherings', 'behavioral_outside_home',
       'behavioral_touch_face', 'doctor_recc_xyz', 'doctor_recc_seasonal',
       'chronic_med_condition', 'child_under_6_months', 'health_worker',
       'health_insurance', 'opinion_xyz_vacc_effective', 'opinion_xyz_risk',
       'opinion_xyz_sick_from_vacc', 'opinion_seas_vacc_effective',
       'opinion_seas_risk', 'opinion_seas_sick_from_vacc', 'age_group',
       'education', 'race', 'sex', 'income_poverty', 'marital_status',
       'rent_or_own', 'employment_status', 'hhs_geo_region', 'census_msa',
       'household_adults', 'household_children', 'employment_industry',
       'employment_occupation'],
      dtype='object')

In [24]:
df.isnull().sum()

respondent_id                      0
xyz_concern                      177
xyz_knowledge                    238
behavioral_antiviral_meds        150
behavioral_avoidance             421
behavioral_face_mask              38
behavioral_wash_hands             82
behavioral_large_gatherings      159
behavioral_outside_home          164
behavioral_touch_face            256
doctor_recc_xyz                 4320
doctor_recc_seasonal            4320
chronic_med_condition           1903
child_under_6_months            1633
health_worker                   1593
health_insurance               24502
opinion_xyz_vacc_effective       789
opinion_xyz_risk                 768
opinion_xyz_sick_from_vacc       770
opinion_seas_vacc_effective      914
opinion_seas_risk               1013
opinion_seas_sick_from_vacc     1058
age_group                          0
education                       2814
race                               0
sex                                0
income_poverty                  8920
m

#### # Insights/Observations:
- The given dataset has some missing values.
- As the columns like health_insurance, employment_industry and employment_occupation has large missing values, we can drop them.

In [25]:
df.drop(["health_insurance", "employment_industry", "employment_occupation"], axis = 1, inplace = True)

In [26]:
df.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...   race     sex             income_poverty  \
0                    1.0  ...  White  Female              Below Poverty   
1                    1.0  ...  White    Male              Below Poverty   
2                    0.0  ...  White    Male  <= $75,000, Above Poverty   
3                    0.0  ...  White  Female              Below Poverty   
4                    1.0  ...  White  Female  <= $75,000, Above Poverty   

   marital_status  rent_or_own   employment_status  hhs_geo_region  \
0     Not Married          Own  Not in Labor Force        oxchjgsf   
1     Not Married         Rent            Employed        bhuqouqj   
2     Not Married          Own            Employed        qufhixun   
3     Not Married         Rent  Not in Labor Force        lrircsnp   
4         Married          Own            Employed        qufhixun   

                 census_msa  household_adults  household_children  
0                   Non-MSA               0.0                 0.0  
1  MSA, Not Principle  City               0.0                 0.0  
2  MSA, Not Principle  City               2.0                 0.0  
3       MSA, Principle City               0.0                 0.0  
4  MSA, Not Principle  City               1.0                 0.0  

[5 rows x 33 columns]

In [27]:
print("df.shape: {}" .format(df.shape))

df.shape: (53415, 33)


In [28]:
df["xyz_concern"].unique()

array([ 1.,  3.,  2.,  0., nan])

In [29]:
df["xyz_concern"].value_counts()

xyz_concern
2.0    21318
1.0    16311
3.0     9114
0.0     6495
Name: count, dtype: int64

In [30]:
df["xyz_knowledge"].unique()

array([ 0.,  2.,  1., nan])

In [31]:
df["xyz_knowledge"].value_counts()

xyz_knowledge
1.0    29227
2.0    19002
0.0     4948
Name: count, dtype: int64

In [32]:
xyz_concern_mode = df["xyz_concern"].mode()
xyz_knowledge_mode = df["xyz_knowledge"].mode()

print("xyz_concern_mode: {}" .format(xyz_concern_mode))
print("xyz_knowledge_mode: {}" .format(xyz_knowledge_mode))

xyz_concern_mode: 0    2.0
Name: xyz_concern, dtype: float64
xyz_knowledge_mode: 0    1.0
Name: xyz_knowledge, dtype: float64


In [33]:
df["xyz_concern"].fillna(xyz_concern_mode[0], inplace = True)
df["xyz_knowledge"].fillna(xyz_knowledge_mode[0], inplace = True)

In [34]:
df[["xyz_concern", "xyz_knowledge"]].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53415 entries, 0 to 53414
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   xyz_concern    53415 non-null  float64
 1   xyz_knowledge  53415 non-null  float64
dtypes: float64(2)
memory usage: 834.7 KB


In [35]:
xyz_concern_mode = df["xyz_concern"].mode()[0]
xyz_knowledge_mode = df["xyz_knowledge"].mode()[0]

print("xyz_concern_mode: {}" .format(xyz_concern_mode))
print("xyz_knowledge_mode: {}" .format(xyz_knowledge_mode))

xyz_concern_mode: 2.0
xyz_knowledge_mode: 1.0


In [36]:
df["xyz_concern"].value_counts()

xyz_concern
2.0    21495
1.0    16311
3.0     9114
0.0     6495
Name: count, dtype: int64

In [37]:
df["xyz_knowledge"].value_counts()

xyz_knowledge
1.0    29465
2.0    19002
0.0     4948
Name: count, dtype: int64

In [38]:
df.isnull().sum()

respondent_id                     0
xyz_concern                       0
xyz_knowledge                     0
behavioral_antiviral_meds       150
behavioral_avoidance            421
behavioral_face_mask             38
behavioral_wash_hands            82
behavioral_large_gatherings     159
behavioral_outside_home         164
behavioral_touch_face           256
doctor_recc_xyz                4320
doctor_recc_seasonal           4320
chronic_med_condition          1903
child_under_6_months           1633
health_worker                  1593
opinion_xyz_vacc_effective      789
opinion_xyz_risk                768
opinion_xyz_sick_from_vacc      770
opinion_seas_vacc_effective     914
opinion_seas_risk              1013
opinion_seas_sick_from_vacc    1058
age_group                         0
education                      2814
race                              0
sex                               0
income_poverty                 8920
marital_status                 2850
rent_or_own                 

In [39]:
df["age_group"].unique()

array(['55 - 64 Years', '35 - 44 Years', '18 - 34 Years', '65+ Years',
       '45 - 54 Years'], dtype=object)

In [40]:
df["age_group"].value_counts()

age_group
65+ Years        13561
55 - 64 Years    11092
45 - 54 Years    10495
18 - 34 Years    10449
35 - 44 Years     7818
Name: count, dtype: int64

In [41]:
df["education"].unique()

array(['< 12 Years', '12 Years', 'College Graduate', 'Some College', nan],
      dtype=object)

In [42]:
df["education"].value_counts()

education
College Graduate    20283
Some College        14123
12 Years            11496
< 12 Years           4699
Name: count, dtype: int64

In [43]:
df["hhs_geo_region"].unique()

array(['oxchjgsf', 'bhuqouqj', 'qufhixun', 'lrircsnp', 'atmpeygn',
       'lzgpxyit', 'fpwskwrf', 'mlyzmhmf', 'dqpwygqj', 'kbazzjca'],
      dtype=object)

In [44]:
df["hhs_geo_region"].value_counts()

hhs_geo_region
lzgpxyit    8660
fpwskwrf    6506
qufhixun    6166
oxchjgsf    5846
bhuqouqj    5713
kbazzjca    5702
mlyzmhmf    4426
lrircsnp    4100
atmpeygn    4019
dqpwygqj    2277
Name: count, dtype: int64

In [45]:
df["census_msa"].unique()

array(['Non-MSA', 'MSA, Not Principle  City', 'MSA, Principle City'],
      dtype=object)

In [46]:
df["census_msa"].value_counts()

census_msa
MSA, Not Principle  City    23571
MSA, Principle City         15669
Non-MSA                     14175
Name: count, dtype: int64

In [47]:
df[["hhs_geo_region", "census_msa"]].isnull().sum()

hhs_geo_region    0
census_msa        0
dtype: int64

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53415 entries, 0 to 53414
Data columns (total 33 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                53415 non-null  int64  
 1   xyz_concern                  53415 non-null  float64
 2   xyz_knowledge                53415 non-null  float64
 3   behavioral_antiviral_meds    53265 non-null  float64
 4   behavioral_avoidance         52994 non-null  float64
 5   behavioral_face_mask         53377 non-null  float64
 6   behavioral_wash_hands        53333 non-null  float64
 7   behavioral_large_gatherings  53256 non-null  float64
 8   behavioral_outside_home      53251 non-null  float64
 9   behavioral_touch_face        53159 non-null  float64
 10  doctor_recc_xyz              49095 non-null  float64
 11  doctor_recc_seasonal         49095 non-null  float64
 12  chronic_med_condition        51512 non-null  float64
 13  child_under_6_mo

In [49]:
df["hhs_geo_region"].dtype

dtype('O')

In [50]:
df["census_msa"].dtype

dtype('O')

In [51]:
for i in df.columns:
    if df[i].isnull().sum() != 0:
        mode_imputation = df[i].mode()[0]
        df[i].fillna(mode_imputation, inplace = True)

In [52]:
df.isnull().sum()

respondent_id                  0
xyz_concern                    0
xyz_knowledge                  0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_xyz                0
doctor_recc_seasonal           0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
opinion_xyz_vacc_effective     0
opinion_xyz_risk               0
opinion_xyz_sick_from_vacc     0
opinion_seas_vacc_effective    0
opinion_seas_risk              0
opinion_seas_sick_from_vacc    0
age_group                      0
education                      0
race                           0
sex                            0
income_poverty                 0
marital_status                 0
rent_or_own                    0
employment_status              0
hhs_geo_region                 0
census_msa

### # Step-02: Check Duplicates

In [53]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
53410    False
53411    False
53412    False
53413    False
53414    False
Length: 53415, dtype: bool

In [54]:
df[df.duplicated() == True]

Empty DataFrame
Columns: [respondent_id, xyz_concern, xyz_knowledge, behavioral_antiviral_meds, behavioral_avoidance, behavioral_face_mask, behavioral_wash_hands, behavioral_large_gatherings, behavioral_outside_home, behavioral_touch_face, doctor_recc_xyz, doctor_recc_seasonal, chronic_med_condition, child_under_6_months, health_worker, opinion_xyz_vacc_effective, opinion_xyz_risk, opinion_xyz_sick_from_vacc, opinion_seas_vacc_effective, opinion_seas_risk, opinion_seas_sick_from_vacc, age_group, education, race, sex, income_poverty, marital_status, rent_or_own, employment_status, hhs_geo_region, census_msa, household_adults, household_children]
Index: []

[0 rows x 33 columns]

#### # Insights/Obseravtion:
- There are zero duplicates in the given dataset.

### # Step-03: Check Datatypes

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53415 entries, 0 to 53414
Data columns (total 33 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                53415 non-null  int64  
 1   xyz_concern                  53415 non-null  float64
 2   xyz_knowledge                53415 non-null  float64
 3   behavioral_antiviral_meds    53415 non-null  float64
 4   behavioral_avoidance         53415 non-null  float64
 5   behavioral_face_mask         53415 non-null  float64
 6   behavioral_wash_hands        53415 non-null  float64
 7   behavioral_large_gatherings  53415 non-null  float64
 8   behavioral_outside_home      53415 non-null  float64
 9   behavioral_touch_face        53415 non-null  float64
 10  doctor_recc_xyz              53415 non-null  float64
 11  doctor_recc_seasonal         53415 non-null  float64
 12  chronic_med_condition        53415 non-null  float64
 13  child_under_6_mo

In [56]:
numerical_feature = [feature for feature in df.columns if df[feature].dtype != "object"]
categorical_feature = [feature for feature in df.columns if df[feature].dtype == "object"]

print("There are {} numerical features: {}" .format(len(numerical_feature), numerical_feature))
print("There are {} categorical feaatures: {}" .format(len(categorical_feature), categorical_feature))

There are 23 numerical features: ['respondent_id', 'xyz_concern', 'xyz_knowledge', 'behavioral_antiviral_meds', 'behavioral_avoidance', 'behavioral_face_mask', 'behavioral_wash_hands', 'behavioral_large_gatherings', 'behavioral_outside_home', 'behavioral_touch_face', 'doctor_recc_xyz', 'doctor_recc_seasonal', 'chronic_med_condition', 'child_under_6_months', 'health_worker', 'opinion_xyz_vacc_effective', 'opinion_xyz_risk', 'opinion_xyz_sick_from_vacc', 'opinion_seas_vacc_effective', 'opinion_seas_risk', 'opinion_seas_sick_from_vacc', 'household_adults', 'household_children']
There are 10 categorical feaatures: ['age_group', 'education', 'race', 'sex', 'income_poverty', 'marital_status', 'rent_or_own', 'employment_status', 'hhs_geo_region', 'census_msa']


### # Step-04: Check the No. of Unique Values

In [57]:
for i in df.columns:
    print("{} : {}" .format(i, df[i].nunique()))

respondent_id : 53415
xyz_concern : 4
xyz_knowledge : 3
behavioral_antiviral_meds : 2
behavioral_avoidance : 2
behavioral_face_mask : 2
behavioral_wash_hands : 2
behavioral_large_gatherings : 2
behavioral_outside_home : 2
behavioral_touch_face : 2
doctor_recc_xyz : 2
doctor_recc_seasonal : 2
chronic_med_condition : 2
child_under_6_months : 2
health_worker : 2
opinion_xyz_vacc_effective : 5
opinion_xyz_risk : 5
opinion_xyz_sick_from_vacc : 5
opinion_seas_vacc_effective : 5
opinion_seas_risk : 5
opinion_seas_sick_from_vacc : 5
age_group : 5
education : 4
race : 4
sex : 2
income_poverty : 3
marital_status : 2
rent_or_own : 2
employment_status : 3
hhs_geo_region : 10
census_msa : 3
household_adults : 4
household_children : 4


#### # Insights/Observations:
##### # No. of Unique Values:
- respondent_id : 53415
- xyz_concern : 4
- xyz_knowledge : 3
- behavioral_antiviral_meds : 2
- behavioral_avoidance : 2
- behavioral_face_mask : 2
- behavioral_wash_hands : 2
- behavioral_large_gatherings : 2
- behavioral_outside_home : 2
- behavioral_touch_face : 2
- doctor_recc_xyz : 2
- doctor_recc_seasonal : 2
- chronic_med_condition : 2
- child_under_6_months : 2
- health_worker : 2
- opinion_xyz_vacc_effective : 5
- opinion_xyz_risk : 5
- opinion_xyz_sick_from_vacc : 5
- opinion_seas_vacc_effective : 5
- opinion_seas_risk : 5
- opinion_seas_sick_from_vacc : 5
- age_group : 5
- education : 4
- race : 4
- sex : 2
- hhs_geo_region : 10
- census_msa : 3
- household_adults : 4
- household_children : 4

### # Step-05: Check Statistics of Dataset

In [58]:
df.describe()

respondent_id   xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
count   53415.000000  53415.000000   53415.000000               53415.000000   
mean    26707.000000      1.622072       1.263110                   0.049106   
std     15419.726651      0.905293       0.615757                   0.216091   
min         0.000000      0.000000       0.000000                   0.000000   
25%     13353.500000      1.000000       1.000000                   0.000000   
50%     26707.000000      2.000000       1.000000                   0.000000   
75%     40060.500000      2.000000       2.000000                   0.000000   
max     53414.000000      3.000000       2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          53415.000000          53415.000000           53415.000000   
mean               0.729851              0.069082               0.826116   
std                0.444041              0.253595               0.379013   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  53415.00000             53415.000000   
mean                       0.35402                 0.336235   
std                        0.47822                 0.472425   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  ...  child_under_6_months  health_worker  \
count           53415.000000  ...          53415.000000   53415.000000   
mean                0.682037  ...              0.081868       0.108378   
std                 0.465689  ...              0.274167       0.310860   
min                 0.000000  ...              0.000000       0.000000   
25%                 0.000000  ...              0.000000       0.000000   
50%                 1.000000  ...              0.000000       0.000000   
75%                 1.000000  ...              0.000000       0.000000   
max                 1.000000  ...              1.000000       1.000000   

       opinion_xyz_vacc_effective  opinion_xyz_risk  \
count                53415.000000      53415.000000   
mean                     3.849874          2.329889   
std                      1.000198          1.271993   
min                      1.000000          1.000000   
25%                      3.000000          1.000000   
50%                      4.000000          2.000000   
75%                      5.000000          4.000000   
max                      5.000000          5.000000   

       opinion_xyz_sick_from_vacc  opinion_seas_vacc_effective  \
count                53415.000000                 53415.000000   
mean                     2.353964                     4.024974   
std                      1.351910                     1.075558   
min                      1.000000                     1.000000   
25%                      1.000000                     4.000000   
50%                      2.000000                     4.000000   
75%                      4.000000                     5.000000   
max                      5.000000                     5.000000   

       opinion_seas_risk  opinion_seas_sick_from_vacc  household_adults  \
count       53415.000000                 53415.000000      53415.000000   
mean            2.700384                     2.108359          0.891379   
std             1.370862                     1.332

In [59]:
df[df.dtypes[df.dtypes == "object"].index].describe()

age_group         education   race     sex             income_poverty  \
count       53415             53415  53415   53415                      53415   
unique          5                 4      4       2                          3   
top     65+ Years  College Graduate  White  Female  <= $75,000, Above Poverty   
freq        13561             23097  42523   31692                      34472   

       marital_status rent_or_own employment_status hhs_geo_region  \
count           53415       53415             53415          53415   
unique              2           2                 3             10   
top           Married         Own          Employed       lzgpxyit   
freq            29953       41535             30126           8660   

                      census_msa  
count                      53415  
unique                         3  
top     MSA, Not Principle  City  
freq                       23571

### # Step-06: Check Various Categories

In [60]:
for i in df.columns:
    if i != "respondent_id" :
        print(df[i].value_counts())

xyz_concern
2.0    21495
1.0    16311
3.0     9114
0.0     6495
Name: count, dtype: int64
xyz_knowledge
1.0    29465
2.0    19002
0.0     4948
Name: count, dtype: int64
behavioral_antiviral_meds
0.0    50792
1.0     2623
Name: count, dtype: int64
behavioral_avoidance
1.0    38985
0.0    14430
Name: count, dtype: int64
behavioral_face_mask
0.0    49725
1.0     3690
Name: count, dtype: int64
behavioral_wash_hands
1.0    44127
0.0     9288
Name: count, dtype: int64
behavioral_large_gatherings
0.0    34505
1.0    18910
Name: count, dtype: int64
behavioral_outside_home
0.0    35455
1.0    17960
Name: count, dtype: int64
behavioral_touch_face
1.0    36431
0.0    16984
Name: count, dtype: int64
doctor_recc_xyz
0.0    42541
1.0    10874
Name: count, dtype: int64
doctor_recc_seasonal
0.0    37133
1.0    16282
Name: count, dtype: int64
chronic_med_condition
0.0    38881
1.0    14534
Name: count, dtype: int64
child_under_6_months
0.0    49042
1.0     4373
Name: count, dtype: int64
health_worker
0

In [61]:
for i in categorical_feature:
    print("no.of unique values : {} \n" .format(df[i].nunique()))
    print("unique values: {} \n" .format(df[i].unique()))
    print("{} \n" .format(df[i].value_counts()))

no.of unique values : 5 

unique values: ['55 - 64 Years' '35 - 44 Years' '18 - 34 Years' '65+ Years'
 '45 - 54 Years'] 

age_group
65+ Years        13561
55 - 64 Years    11092
45 - 54 Years    10495
18 - 34 Years    10449
35 - 44 Years     7818
Name: count, dtype: int64 

no.of unique values : 4 

unique values: ['< 12 Years' '12 Years' 'College Graduate' 'Some College'] 

education
College Graduate    23097
Some College        14123
12 Years            11496
< 12 Years           4699
Name: count, dtype: int64 

no.of unique values : 4 

unique values: ['White' 'Black' 'Other or Multiple' 'Hispanic'] 

race
White                42523
Black                 4292
Hispanic              3458
Other or Multiple     3142
Name: count, dtype: int64 

no.of unique values : 2 

unique values: ['Female' 'Male'] 

sex
Female    31692
Male      21723
Name: count, dtype: int64 

no.of unique values : 3 

unique values: ['Below Poverty' '<= $75,000, Above Poverty' '> $75,000'] 

income_poverty
<= $75

### # Feature Engineering - Encoding

In [62]:
pip install -U scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [63]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import LabelEncoder

In [64]:
nominal_encoder = OneHotEncoder()
ordinal_encoder = OrdinalEncoder()
label_encoder = LabelEncoder()

In [65]:
nominal_encoder

OneHotEncoder()

In [66]:
ordinal_encoder

OrdinalEncoder()

In [67]:
label_encoder

LabelEncoder()

In [68]:
df["sex"] = np.where(df["sex"] == "Female", 0, 1)
df["marital_status"] = np.where(df["marital_status"] == "Not Married", 0, 1)
df["rent_or_own"] = np.where(df["rent_or_own"] == "Rent", 0, 1)

In [69]:
df[df.dtypes[df.dtypes == "object"].index]

age_group         education   race             income_poverty  \
0      55 - 64 Years        < 12 Years  White              Below Poverty   
1      35 - 44 Years          12 Years  White              Below Poverty   
2      18 - 34 Years  College Graduate  White  <= $75,000, Above Poverty   
3          65+ Years          12 Years  White              Below Poverty   
4      45 - 54 Years      Some College  White  <= $75,000, Above Poverty   
...              ...               ...    ...                        ...   
53410  35 - 44 Years  College Graduate  White  <= $75,000, Above Poverty   
53411  18 - 34 Years          12 Years  White              Below Poverty   
53412  18 - 34 Years      Some College  White              Below Poverty   
53413  55 - 64 Years      Some College  White  <= $75,000, Above Poverty   
53414  45 - 54 Years  College Graduate  White  <= $75,000, Above Poverty   

        employment_status hhs_geo_region                census_msa  
0      Not in Labor Force       oxchjgsf                   Non-MSA  
1                Employed       bhuqouqj  MSA, Not Principle  City  
2                Employed       qufhixun  MSA, Not Principle  City  
3      Not in Labor Force       lrircsnp       MSA, Principle City  
4                Employed       qufhixun  MSA, Not Principle  City  
...                   ...            ...                       ...  
53410            Employed       dqpwygqj       MSA, Principle City  
53411            Employed       qufhixun                   Non-MSA  
53412  Not in Labor Force       qufhixun  MSA, Not Principle  City  
53413  Not in Labor Force       bhuqouqj  MSA, Not Principle  City  
53414            Employed       lrircsnp       MSA, Principle City  

[53415 rows x 7 columns]

In [70]:
df["income_poverty"].unique()

array(['Below Poverty', '<= $75,000, Above Poverty', '> $75,000'],
      dtype=object)

In [71]:
ordinal_encoder = OrdinalEncoder(categories = [list(df["income_poverty"].unique())])

In [72]:
ordinal_encoder

OrdinalEncoder(categories=[['Below Poverty', '<= $75,000, Above Poverty',
                            '> $75,000']])

In [73]:
ordinal_encoded = ordinal_encoder.fit_transform(df[["income_poverty"]])

In [74]:
ordinal_encoded

array([[0.],
       [0.],
       [1.],
       ...,
       [0.],
       [1.],
       [1.]])

In [75]:
df_income_poverty = pd.DataFrame(ordinal_encoded, columns = ["income_poverty"])

In [76]:
df["employment_status"].unique()

array(['Not in Labor Force', 'Employed', 'Unemployed'], dtype=object)

In [77]:
ordinal_encoder = OrdinalEncoder(categories = [['Unemployed', 'Not in Labor Force', 'Employed']])

In [78]:
ordinal_encoder

OrdinalEncoder(categories=[['Unemployed', 'Not in Labor Force', 'Employed']])

In [79]:
ordinal_encoded = ordinal_encoder.fit_transform(df[["employment_status"]])

In [80]:
ordinal_encoded

array([[1.],
       [2.],
       [2.],
       ...,
       [1.],
       [1.],
       [2.]])

In [81]:
df_employment_status = pd.DataFrame(ordinal_encoded, columns = ["employment_status"])

In [82]:
df["census_msa"].unique()

array(['Non-MSA', 'MSA, Not Principle  City', 'MSA, Principle City'],
      dtype=object)

In [83]:
ordinal_encoder = OrdinalEncoder(categories = [list(df["census_msa"].unique())])

In [84]:
ordinal_encoder

OrdinalEncoder(categories=[['Non-MSA', 'MSA, Not Principle  City',
                            'MSA, Principle City']])

In [85]:
ordinal_encoded = ordinal_encoder.fit_transform(df[["census_msa"]])

In [86]:
ordinal_encoded

array([[0.],
       [1.],
       [1.],
       ...,
       [1.],
       [1.],
       [2.]])

In [87]:
df_census_msa = pd.DataFrame(ordinal_encoded, columns = ["census_msa"])

In [88]:
df.drop(["income_poverty", "employment_status", "census_msa"], axis = 1, inplace = True)

In [89]:
df = pd.concat([df, df_income_poverty, df_employment_status, df_census_msa], axis = 1)

In [90]:
df.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...   race  sex  marital_status  rent_or_own  \
0                    1.0  ...  White    0               0            1   
1                    1.0  ...  White    1               0            0   
2                    0.0  ...  White    1               0            1   
3                    0.0  ...  White    0               0            0   
4                    1.0  ...  White    0               1            1   

   hhs_geo_region  household_adults  household_children  income_poverty  \
0        oxchjgsf               0.0                 0.0             0.0   
1        bhuqouqj               0.0                 0.0             0.0   
2        qufhixun               2.0                 0.0             1.0   
3        lrircsnp               0.0                 0.0             0.0   
4        qufhixun               1.0                 0.0             1.0   

   employment_status  census_msa  
0                1.0         0.0  
1                2.0         1.0  
2                2.0         1.0  
3                1.0         2.0  
4                2.0         1.0  

[5 rows x 33 columns]

In [91]:
df["education"].unique()

array(['< 12 Years', '12 Years', 'College Graduate', 'Some College'],
      dtype=object)

In [92]:
ordinal_encoder = OrdinalEncoder(categories = [['< 12 Years', '12 Years', 'Some College', 'College Graduate']])

In [93]:
ordinal_encoder

OrdinalEncoder(categories=[['< 12 Years', '12 Years', 'Some College',
                            'College Graduate']])

In [94]:
ordinal_encoded = ordinal_encoder.fit_transform(df[["education"]])

In [95]:
ordinal_encoded

array([[0.],
       [1.],
       [3.],
       ...,
       [2.],
       [2.],
       [3.]])

In [96]:
df_education = pd.DataFrame(ordinal_encoded, columns = ["education"])

In [97]:
df["age_group"].unique()

array(['55 - 64 Years', '35 - 44 Years', '18 - 34 Years', '65+ Years',
       '45 - 54 Years'], dtype=object)

In [98]:
ordinal_encoder = OrdinalEncoder(categories = [['18 - 34 Years', '35 - 44 Years', '45 - 54 Years', '55 - 64 Years', '65+ Years']])

In [99]:
ordinal_encoded = ordinal_encoder.fit_transform(df[["age_group"]])

In [100]:
ordinal_encoded

array([[3.],
       [1.],
       [0.],
       ...,
       [0.],
       [3.],
       [2.]])

In [101]:
df_age_group = pd.DataFrame(ordinal_encoded, columns = ["age_group"])

In [102]:
df.drop(["age_group", "education"], axis = 1, inplace = True)

In [103]:
df = pd.concat([df, df_age_group, df_education], axis = 1)

In [104]:
df.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...  marital_status  rent_or_own  hhs_geo_region  \
0                    1.0  ...               0            1        oxchjgsf   
1                    1.0  ...               0            0        bhuqouqj   
2                    0.0  ...               0            1        qufhixun   
3                    0.0  ...               0            0        lrircsnp   
4                    1.0  ...               1            1        qufhixun   

   household_adults  household_children  income_poverty  employment_status  \
0               0.0                 0.0             0.0                1.0   
1               0.0                 0.0             0.0                2.0   
2               2.0                 0.0             1.0                2.0   
3               0.0                 0.0             0.0                1.0   
4               1.0                 0.0             1.0                2.0   

   census_msa  age_group  education  
0         0.0        3.0        0.0  
1         1.0        1.0        1.0  
2         1.0        0.0        3.0  
3         2.0        4.0        1.0  
4         1.0        2.0        2.0  

[5 rows x 33 columns]

In [105]:
df["race"].unique()

array(['White', 'Black', 'Other or Multiple', 'Hispanic'], dtype=object)

In [106]:
label_encoder = LabelEncoder()

In [107]:
label_encoded = label_encoder.fit_transform(df["race"])

In [108]:
label_encoded

array([3, 3, 3, ..., 3, 3, 3])

In [109]:
df_race = pd.DataFrame(label_encoded, columns = ["race"])

In [110]:
label_encoder = LabelEncoder()

In [111]:
label_encoded = label_encoder.fit_transform(df["hhs_geo_region"])

In [112]:
label_encoded

array([8, 1, 9, ..., 9, 1, 5])

In [113]:
df_hhs_geo_region = pd.DataFrame(label_encoded, columns = ["hhs_geo_region"])

In [114]:
df.drop(["race", "hhs_geo_region"], axis = 1, inplace = True)

In [115]:
df = pd.concat([df, df_race, df_hhs_geo_region], axis = 1)

In [116]:
df.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...  rent_or_own  household_adults  \
0                    1.0  ...            1               0.0   
1                    1.0  ...            0               0.0   
2                    0.0  ...            1               2.0   
3                    0.0  ...            0               0.0   
4                    1.0  ...            1               1.0   

   household_children  income_poverty  employment_status  census_msa  \
0                 0.0             0.0                1.0         0.0   
1                 0.0             0.0                2.0         1.0   
2                 0.0             1.0                2.0         1.0   
3                 0.0             0.0                1.0         2.0   
4                 0.0             1.0                2.0         1.0   

   age_group  education  race  hhs_geo_region  
0        3.0        0.0     3               8  
1        1.0        1.0     3               1  
2        0.0        3.0     3               9  
3        4.0        1.0     3               5  
4        2.0        2.0     3               9  

[5 rows x 33 columns]

In [117]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53415 entries, 0 to 53414
Data columns (total 33 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   respondent_id                53415 non-null  int64  
 1   xyz_concern                  53415 non-null  float64
 2   xyz_knowledge                53415 non-null  float64
 3   behavioral_antiviral_meds    53415 non-null  float64
 4   behavioral_avoidance         53415 non-null  float64
 5   behavioral_face_mask         53415 non-null  float64
 6   behavioral_wash_hands        53415 non-null  float64
 7   behavioral_large_gatherings  53415 non-null  float64
 8   behavioral_outside_home      53415 non-null  float64
 9   behavioral_touch_face        53415 non-null  float64
 10  doctor_recc_xyz              53415 non-null  float64
 11  doctor_recc_seasonal         53415 non-null  float64
 12  chronic_med_condition        53415 non-null  float64
 13  child_under_6_mo

### # Logistic Regression - Multilabel Classsification

In [123]:
x_train = df.iloc[:26707].reset_index(drop = True)
x_test = df.iloc[26707:].reset_index(drop = True)

In [124]:
x_train.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...  rent_or_own  household_adults  \
0                    1.0  ...            1               0.0   
1                    1.0  ...            0               0.0   
2                    0.0  ...            1               2.0   
3                    0.0  ...            0               0.0   
4                    1.0  ...            1               1.0   

   household_children  income_poverty  employment_status  census_msa  \
0                 0.0             0.0                1.0         0.0   
1                 0.0             0.0                2.0         1.0   
2                 0.0             1.0                2.0         1.0   
3                 0.0             0.0                1.0         2.0   
4                 0.0             1.0                2.0         1.0   

   age_group  education  race  hhs_geo_region  
0        3.0        0.0     3               8  
1        1.0        1.0     3               1  
2        0.0        3.0     3               9  
3        4.0        1.0     3               5  
4        2.0        2.0     3               9  

[5 rows x 33 columns]

In [125]:
x_test.head()

respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0          26707          2.0            2.0                        0.0   
1          26708          1.0            1.0                        0.0   
2          26709          2.0            2.0                        0.0   
3          26710          1.0            1.0                        0.0   
4          26711          3.0            1.0                        1.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   1.0                   0.0                    1.0   
1                   0.0                   0.0                    0.0   
2                   0.0                   1.0                    1.0   
3                   0.0                   0.0                    0.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          1.0                      0.0   
1                          0.0                      0.0   
2                          1.0                      1.0   
3                          0.0                      0.0   
4                          1.0                      1.0   

   behavioral_touch_face  ...  rent_or_own  household_adults  \
0                    1.0  ...            0               1.0   
1                    0.0  ...            0               3.0   
2                    1.0  ...            1               1.0   
3                    0.0  ...            1               1.0   
4                    1.0  ...            1               0.0   

   household_children  income_poverty  employment_status  census_msa  \
0                 0.0             2.0                2.0         1.0   
1                 0.0             0.0                2.0         0.0   
2                 0.0             2.0                2.0         0.0   
3                 0.0             1.0                1.0         1.0   
4                 1.0             1.0                2.0         0.0   

   age_group  education  race  hhs_geo_region  
0        1.0        3.0     1               7  
1        0.0        1.0     3               1  
2        3.0        3.0     3               5  
3        4.0        1.0     3               5  
4        1.0        1.0     0               6  

[5 rows x 33 columns]

In [126]:
y_train

respondent_id  xyz_vaccine  seasonal_vaccine
0                  0            0                 0
1                  1            0                 1
2                  2            0                 0
3                  3            0                 1
4                  4            0                 0
...              ...          ...               ...
26702          26702            0                 0
26703          26703            0                 0
26704          26704            0                 1
26705          26705            0                 0
26706          26706            0                 0

[26707 rows x 3 columns]

In [127]:
y_train.drop("respondent_id", axis = 1, inplace = True)

In [128]:
y_train.head()

xyz_vaccine  seasonal_vaccine
0            0                 0
1            0                 1
2            0                 0
3            0                 1
4            0                 0

In [131]:

from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.multioutput import MultiOutputClassifier

In [147]:
logistic_regression = LogisticRegression(max_iter = 10000)
multioutput_classifier = MultiOutputClassifier(logistic_regression)

In [148]:
multioutput_classifier.fit(x_train, y_train)

MultiOutputClassifier(estimator=LogisticRegression(max_iter=10000))

In [149]:
y_pred = multioutput_classifier.predict(x_test)

In [150]:
y_pred

array([[0, 0],
       [0, 0],
       [0, 1],
       ...,
       [0, 0],
       [0, 0],
       [1, 1]], dtype=int64)

In [151]:
y_pred_proba = multioutput_classifier.predict_proba(x_test)

In [166]:
y_01 = np.matrix(y_pred_proba[0])

In [167]:
y_01

matrix([[0.92136444, 0.07863556],
        [0.95729227, 0.04270773],
        [0.55404841, 0.44595159],
        ...,
        [0.87031115, 0.12968885],
        [0.94459881, 0.05540119],
        [0.43068145, 0.56931855]])

In [168]:
y_02 = np.matrix(y_pred_proba[1])

In [169]:
y_02

matrix([[0.78222868, 0.21777132],
        [0.953583  , 0.046417  ],
        [0.28181313, 0.71818687],
        ...,
        [0.83589743, 0.16410257],
        [0.60380197, 0.39619803],
        [0.43788964, 0.56211036]])

In [155]:
y_submit = pd.DataFrame(x_test[["respondent_id"]])

In [156]:
y_submit

respondent_id
0              26707
1              26708
2              26709
3              26710
4              26711
...              ...
26703          53410
26704          53411
26705          53412
26706          53413
26707          53414

[26708 rows x 1 columns]

In [173]:
y_xyz_vaccine = pd.DataFrame(y_01, columns = ["not_received", "received"])

In [174]:
y_xyz_vaccine

not_received  received
0          0.921364  0.078636
1          0.957292  0.042708
2          0.554048  0.445952
3          0.523205  0.476795
4          0.795926  0.204074
...             ...       ...
26703      0.662019  0.337981
26704      0.920447  0.079553
26705      0.870311  0.129689
26706      0.944599  0.055401
26707      0.430681  0.569319

[26708 rows x 2 columns]

In [175]:
y_seasonal_vaccine = pd.DataFrame(y_02, columns = ["not_received", "received"])

In [176]:
y_seasonal_vaccine

not_received  received
0          0.782229  0.217771
1          0.953583  0.046417
2          0.281813  0.718187
3          0.136602  0.863398
4          0.539068  0.460932
...             ...       ...
26703      0.484226  0.515774
26704      0.745633  0.254367
26705      0.835897  0.164103
26706      0.603802  0.396198
26707      0.437890  0.562110

[26708 rows x 2 columns]

In [178]:
y_submit = pd.concat([y_submit, y_xyz_vaccine["received"], y_seasonal_vaccine["received"]], axis = 1)

In [179]:
y_submit

respondent_id  received  received
0              26707  0.078636  0.217771
1              26708  0.042708  0.046417
2              26709  0.445952  0.718187
3              26710  0.476795  0.863398
4              26711  0.204074  0.460932
...              ...       ...       ...
26703          53410  0.337981  0.515774
26704          53411  0.079553  0.254367
26705          53412  0.129689  0.164103
26706          53413  0.055401  0.396198
26707          53414  0.569319  0.562110

[26708 rows x 3 columns]

In [180]:
y_submit.to_csv("submission_file.csv")